### ISSUE TITLE
No guards for frozen dataclasses and readonly configs

### ISSUE THREAD

**Issue by vmoens** [pytorch/pytorch#138226]
### 🚀 The feature, motivation and pitch

Currently, a guard will be put for a frozen dataclass:
```
TORCH_LOGS="+guards" python -c """from dataclasses import dataclass
import torch

@dataclass(frozen=True)
class MyDC:
    x: int

@torch.compile(fullgraph=True)
def func(dc: MyDC, t):
    return dc.x + t

func(MyDC(0), torch.randn(()))
"""
```
gives a guard for
```
V1017 16:30:38.479000 75647 torch/_dynamo/guards.py:2314] [0/0] [__guards] | | +- GuardManager: source=L['dc'].x, accessed_by=GetAttrGuardAccessor(x)
V1017 16:30:38.479000 75647 torch/_dynamo/guards.py:2314] [0/0] [__guards] | | | +- EQUALS_MATCH: L['dc'].x == 0                                                #   # <string>:10 in func
```

This is unnecessary as `dc.x` cannot be changed (`dc` can be guarded but `dc.x` is frozen).

The same would apply for readonly configs from omegaconf (when #138224 is fixed, this code will show guards for x):
```python
import torch
from omegaconf import DictConfig

@torch.compile(fullgraph=True) # this will break currently
def func(dc, t):
    return dc.x + t

cfg = DictConfig({'x': 0}, flags={'readonly': True})
func(cfg, torch.randn(()))
```

### Alternatives

I guess someone could hack their way through a frozen dataclass and that could lead to some errors but is it that much of a risk?

**Comment by anijain2305** [comment]
After thinking more, we still need a guard. For the same function, you could still pass a new config or a new frozen dataclass that has different values from the one that triggered the compilation. 

But there is one optimization idea here - if a function argument was `frozen dataclass` and the `id` of the argument is same at runtime, then we can skip the guards. This is theoretically possible, I need to think more how to implement this.

**Comment by ezyang** [comment]
It's pretty similar to dict tag optimization

**Comment by anijain2305** [comment]
Looked little bit more into this. Even if the dataclass is frozen, you can still have a mutable attribute on the dataclass (like a list) that you can mutate w/o any issues. So, we have to recurse the dataclass attributes for mutable objects any ways.

For frozen dataclasses attribute guarding, there are 2 parts
1) Accessing the attribute itself - like dc.x
2) Guarding on the attribute like - dc.x == 0

The good news it that we already short-circuit (2) if the object was immutable and `id` of attribute is same as the guarded one.  But, we still spend time on (1). For frozen dataclasses, we can try to optimize away this access perhaps. I am not sure how much we will win but I will try to implement that.


For omegaconf DictConfig, the store is entirely different. Dynamo traces through the omegaconf python code, which is a complicated datastructure and inserts a large number of guards on the internals of the DictConfig object. All these guards seem necessary, and there is no easy way to tell which ones can be skipped.

**Comment by StrongerXi** [comment]
> we already short-circuit (2) if the object was immutable and id of attribute is same as the guarded one.

By "short-circuit" are you referring to the dict-tag optimization, or something else? From the code it doesn't look like we'll optimize?
https://github.com/pytorch/pytorch/blob/f30f43f594374aa7d4578bd8c7f69a1f7b14ad8e/torch/csrc/dynamo/guards.cpp#L2777-L2784

> For frozen dataclasses, we can try to optimize away this access perhaps. I am not sure how much we will win but I will try to implement that.

#141061 optimizes away both (1) and (2), but yeah I'm not sure about the gain. We could probably lazy evaluate this problem until it's actually concerning, just to avoid adding more complexity to the code.

**Comment by anijain2305** [comment]
Yes, I was talking about the dict-tag optimization.  For frozen dataclasses, instead of creating GetAttrGuardAccessor on something like dc.x, we should do GetGenericDictAccessor following by DictGetItemAccessor. This will directly use dict-tag optimization and should give benefit.

**Comment by StrongerXi** [comment]
> For frozen dataclasses, instead of creating GetAttrGuardAccessor on something like dc.x, we should do GetGenericDictAccessor following by DictGetItemAccessor. This will directly use dict-tag optimization and should give benefit.

This is simpler and can be done in Python, but I'll hold off it until I finish some other optimization for dict-tag.

**Comment by Xynonners** [comment]
hello,
I'm wondering if there's just a way to mark an object as immutable (a dict that I know will not change).

**Comment by ezyang** [comment]
if you like living on the dangerous side check out https://github.com/pytorch/pytorch/pull/140251

**Comment by Xynonners** [comment]
well it's just one global dict object, though I definitely still need guards to work since the function has a ton of control flow

**Comment by Xynonners** [comment]
ideally it'd be like a mark_static for objects

**Comment by ezyang** [comment]
I think it wouldn't be too difficult to have a magic attribute on objects that have `__dict__` to cause them to be treated as readonly, or potentially some sort of type test. So I guess the problem is persuading the Dynamo developers that this is a good API to have.

**Comment by Xynonners** [comment]
I think a more general solution might just be to have a way to mark an object as a global constant?

**Comment by anijain2305** [comment]
So I made the dict guards more lazy. Can you check if on main branch, you see fewer guards for your dict object? 

If you still see large number of guards, please share a repro and I can see if I can insert only the necessary guards.

**Comment by masnesral** [comment]
[issue scrubbing] @StrongerXi this is assigned to use, but Animesh's last update kinda indicates this is probably fixed. Close it?

**Comment by StrongerXi** [comment]
Let's leave this open for now. Animesh's optimization might mitigate some of it, but I think the fundamental ask here is a way for users to tell `torch.compile` "hey I promise this object will be constant despite python dynamicity". We might also see more use of these config things as we integrate `torch.compile` with other frameworks.

Also, it sounds like there have been similar asks for a guard-skipping mechanism from folks, e.g., vLLM? @anijain2305 @zou3519
